File: baekjoon/단계별로 풀어보기/스택/4949_균형잡힌 세상.py
# https://www.acmicpc.net/problem/4949

import re
from sys import stdin

input = stdin.readline
while True:
    s = input().rstrip()
    if s == '.':
        break
    s = re.sub(r"[^\[\]\(\)]", "", s)
    while ('()' in s) or ('[]' in s):
        s = s.replace('()', '')
        s = s.replace('[]', '')
    print('no' if s else 'yes')

File: C#/Baekjoon/스택/4949_균형잡힌 세상.cs
//https://www.acmicpc.net/problem/4949

using System.Text.RegularExpressions;

using var sr = new StreamReader(new BufferedStream(Console.OpenStandardInput()));
using var sw = new StreamWriter(new BufferedStream(Console.OpenStandardOutput()));
string l;
char m;
while (true)
{
    var s = new Stack<char>();
    l = sr.ReadLine();
    if (l == ".") break;
    l = Regex.Replace(l, @"[^\[\]()]", "");
    foreach (char c in l)
    {
        if (s.TryPeek(out m) && (m == '(' && c == ')' || m == '[' && c == ']')) s.Pop();
        else s.Push(c);
    }
    sw.WriteLine(s.Count == 0 ? "yes" : "no");
}